### PR TITLE
unrecognized hex char fix

### PR DESCRIPTION
**Pull request by suskumar-MSFT**
Each EscapeChar (%) must be followed by two integer values if that is not the case then it is likely that the truncation left incomplete escapes. So making a fix to remove if the string ends with % or % followed by a single digit.

### CODE REVIEW

File: src/PAModel/Utility/Utilities.cs [hunk 1]
```diff
@@ -375,11 +375,25 @@ public static string TruncateNameIfTooLong(string name)
...
         }
 
+        /// <summary>
+        /// Truncates a string with the given length and strips off incomplete escape characters if any.
+        /// Each EscapeChar (%) must be followed by two integer values if that is not the case then it is likely that the truncation left incomplete escapes.
+        /// </summary>
+        /// <param name="name">The string to be truncated</param>
+        /// <param name="length">The max length of the truncated string.</param>
+        /// <returns>Truncated string.</returns>
+        private static string TruncateName(string name, int length)
```
Comment (MikeStall) [inline src/PAModel/Utility/Utilities.cs:391]: >TruncateName [](http://example.com/codeflow?start=30&length=12)

Is there some unit test to add?

File: src/PAModel/Utility/Utilities.cs [hunk 2]
```diff
@@ -375,11 +375,25 @@ public static string TruncateNameIfTooLong(string name)
...
+        {
+            int removeTrailingCharsLength = name[length - 1] == EscapeChar ? 1 : name[length - 2] == EscapeChar ? 2 : 0;
```
Comment (lesaltzm) [inline src/PAModel/Utility/Utilities.cs:393]: Very minor nit, but adding parens would make this nested ternary much more readable.

File: src/PAModel/Utility/Utilities.cs [hunk 3]
```diff
@@ -375,11 +375,25 @@ public static string TruncateNameIfTooLong(string name)
             {
                 // limit the hash to 3 characters by doing a module by 4096 (16^3)
                 var hash = (GetHash(escapedName) % 4096).ToString("x3");
-                escapedName = escapedName.Substring(0, MaxNameLength - (hash.Length + 1)) + "_" + hash;
+                escapedName = TruncateName(escapedName, MaxNameLength - hash.Length -1) + "_" + hash;
+
```
Comment (lesaltzm) [inline src/PAModel/Utility/Utilities.cs:379]: nit: extra newline

File: src/PAModel/Utility/Utilities.cs (hunk 2 above)
Comment (MikeStall) [inline src/PAModel/Utility/Utilities.cs:393]: >name[length - 1] == EscapeChar ? 1 : name[length - 2] == EscapeChar ? 2 : 0; [](http://example.com/codeflow?start=44&length=76)

nit: can you put parenthesis in here to disambiguate?

File: src/PAModel/Utility/Utilities.cs (hunk 1 above)
Comment (suskumar-MSFT) [reply src/PAModel/Utility/Utilities.cs:390]: There was an existing unit test which I have updated as needed

File: src/PAModel/Utility/Utilities.cs [hunk 4]
```diff
@@ -375,11 +375,24 @@ public static string TruncateNameIfTooLong(string name)
...
+        /// <summary>
+        /// Truncates a string with the given length and strips off incomplete escape characters if any.
+        /// Each EscapeChar (%) must be followed by two integer values if that is not the case then it is likely that the truncation left incomplete escapes.
+        /// </summary>
+        /// <param name="name">The string to be truncated</param>
+        /// <param name="length">The max length of the truncated string.</param>
+        /// <returns>Truncated string.</returns>
+        private static string TruncateName(string name, int length)
+        {
+            int removeTrailingCharsLength = name[length - 1] == EscapeChar ? 1 : (name[length - 2] == EscapeChar ? 2 : 0);
```
Comment (MikeStall) [inline src/PAModel/Utility/Utilities.cs:392]: >name[length - 1]  [](http://example.com/codeflow?start=44&length=17)

what if name is just 1 char?

Comment (suskumar-MSFT) [reply src/PAModel/Utility/Utilities.cs:392]: currently this only gets invoked if the filename is longer than 50 chars

Comment (MikeStall) [reply src/PAModel/Utility/Utilities.cs:392]: resolved offline: check on line 374 protects us.

**Review by MikeStall — APPROVED** [review]
:shipit:

**Pull request merged**